text/3605-freeze-uninit.md

```diff
@@ -0,0 +1,219 @@
+    * Along the same lines, a possible fast floating-point operation set that yields uninit on invalid (such as NaN or Infinite) results, stored as `MaybeUninit`, then frozen upon return as `f32`/`f64`.
+    * Note that such operations require compiler support, and these operations are *not* defined by this RFC.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Two new library interfaces are defined for unsafe code users:
+
+```rust
+// in module `core::ptr`
+pub unsafe fn read_freeze<T>(ptr: *const T) -> T;
+
+impl<T> *const T{
+    pub unsafe fn read_freeze(self) -> T;
+}
+impl<T> *mut T{
+    pub unsafe fn read_freeze(self) -> T;
+}
+
+// in module `core::mem`
+impl<T> MaybeUninit<T>{
+    pub fn freeze(self) -> Self;
+}
+```
+
+`read_freeze` is an operation that loads a number of bytes from (potentially uninitialized) memory, replacing every byte in the source that is uninitialized with an arbitrary-but-initialized value.  
+The function has similar safety constraints to `core::ptr::read`, in that the source pointer must be *dereferenceable* and well aligned, as well as pointing to a valid value of type `T` (after replacing uninitialized bytes).  
+The same-name functions on the raw pointer types are identical to the free function version, and are provided for convience. 
+Only the bytes of the return value are frozen. The bytes behind the pointer argument are unmodified.  
+**The `read_freeze` operation does not freeze any padding bytes of `T` (if any are present), and those are set to uninit after the read as usual.**
+
+`MaybeUninit::freeze` is a safe, by-value version of `read_freeze`. It takes in a `MaybeUninit` and yields an initialized value, which is either exactly `self` if it is already initialized, or some arbitrary value if it is not. 
+
+The result of `core::ptr::read_freeze` or `MaybeUninit::freeze` is not guaranteed to be valid for `T` if `T` has a more complex validity invariant than an array of `u8` (for example, `char`, `bool`, or a reference type). 
+However, the result is guaranteed to be valid for an integer or floating point type, or an aggregate (struct, union, tuple, or array) only containing (recursively) those types. 
+
+Note that frozen bytes are arbitrary, not random. 
+Rust code must not rely on frozen uninitialized bytes (or unfrozen uninitialized bytes) as a source of entropy, only as values that it does not care about and is fine with garbage-but-consistent computational results from. 
+
+For example, the following function:
+```rust
+pub fn gen_random() -> i32{
+    unsafe{MaybeUnit::uninit().freeze().assume_init()}
+}
+```
+
+can be validily optimized to:
+```rust
+pub fn gen_random() -> i32{
+    4 // Compiler chose a constant by fair dice roll
+}
+```
+
+(See also [XKCD 221](https://xkcd.com/221/))
+
+Note that the value `4` was chosen for expository purposes only, and the same optimization could be validly replace by any other constant, or not at all.
```

Comment:
it can also be replaced with variables from functions it's inlined into.

Reply:
It can be replaced by basically anything, the RFC does not and should not try to have an exhaustive list.